### ISSUE TITLE
Unable to open some.json output in chrome://tracing

### ISSUE THREAD

**Issue by ASoldo** [omaskery/teffy#10]
Hi, I'm having a problem opening this json file in chrome://tracing.

I have used snippet from docs to run and generate this json output:
```go
func tracer() {
	t, _ := trace.TraceToFile("some.json")
	defer t.Close()

	defer t.BeginDuration("my event", trace.WithCategories("cool", "categories")).End()

	// your amazing code
	t.Instant("wow a thing happened!", trace.WithStackTrace())
}
```

```json
[
  {
    "ph": "B",
    "name": "my event",
    "cat": "cool,categories",
    "ts": 1663559685039795,
    "pid": 8452
  },
  {
    "ph": "I",
    "name": "wow a thing happened!",
    "ts": 1663559685040315,
    "pid": 8452,
    "stack": [
      "github.com/omaskery/teffy/pkg/util/trace.(*Tracer).Instant:213",
      "main.tracer:39",
      "main.main:28",
      "runtime.main:250",
      "runtime.goexit:1594"
    ],
    "s": "t"
  },
  {
    "ph": "E",
    "name": "my event",
    "ts": 1663559685040315,
    "pid": 8452
  }
]
``` 

This is the error message I got: 
```
While importing:
TypeError: this.model_.resolveStackToStackFrame_ is not a function
    at TraceEventImporter.getStackFrameForEvent_ (chrome://tracing/tracing.js:6354:42)
    at TraceEventImporter.processDurationEvent (chrome://tracing/tracing.js:6283:346)
    at TraceEventImporter.processInstantEvent (chrome://tracing/tracing.js:6295:45)
    at TraceEventImporter.processEvent_ (chrome://tracing/tracing.js:6337:312)
    at TraceEventImporter.importEvents (chrome://tracing/tracing.js:6329:70)
    at chrome://tracing/tracing.js:2018:232
    at chrome://tracing/tracing.js:2014:268
    at Task.run (chrome://tracing/tracing.js:3259:95)
    at runAnother (chrome://tracing/tracing.js:3262:371)
    at runTask (chrome://tracing/tracing.js:3026:57)
```
Go version: go 1.19

Can you help me with this? Am I missing something? 
Thank you!

**Comment by omaskery** [comment]
I can confirm this issue, however from what I can tell I am complying with the TEF documentation ([explanation of `"stack"` key](https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/preview#heading=h.yr703knxre9f), and [specific suggestion that the `"stack"` key can be populated in `Instant` events](https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/preview#heading=h.lenwiilchoxp)).

There is a remark in the `Instant` documentation that I've linked above that reads (emphasis mine):

> Thread-scoped events can have stack traces associated with them in the [same style as duration events](https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/preview#heading=h.yr703knxre9f), by putting either sf or stack records on the event. **Process-scoped and global-scoped events do not support stack traces at this time.**

It could be that my example is wrong, and I am creating a process/global scoped event with a stack trace, which is unsupported. However, googling the error message leads me to this issue on the source code for the chrome tracing window: https://github.com/catapult-project/catapult/issues/3996

As a result, I'm a little unclear whether this is actually an issue with the chrome tracing window, or whether my example is wrong. What are your thoughts? It feels to me like the trace viewer just doesn't support that feature yet?

---

As a short-term workaround, if this is blocking you: I believe you can simply omit the `trace.WithStackTrace()` argument to the `Instant` call.

**Comment by ASoldo** [comment]
I really don't know... but then I found out about "runtime/trace" and that output can be previewed with go tools trace trace.out.

This is the function I used: 
```go
func tracer() {
	f, err := os.Create("trace.out")
	if err != nil {
		panic(err)
	}
	defer f.Close()
	err = trace.Start(f)
	if err != nil {
		panic(err)
	}
	defer trace.Stop()

	fmt.Println("Tracing")
}
```

**Comment by omaskery** [comment]
Are you able to export the resulting Trace Event Format JSON from the viewer? It would be interesting to see what it does.

**Comment by ASoldo** [comment]
no, I didn't see any export option. Go cli is making a binary file that cannot be previewed but as I see they parse it and splitting it.
```
2022/09/19 14:51:49 Parsing trace...
2022/09/19 14:51:49 Splitting trace...
2022/09/19 14:51:49 Opening browser. Trace viewer is listening on http://127.0.0.1:53843
```

**Comment by omaskery** [comment]
Is the trace viewer not the one built into chrome, which has a save button? I was wondering if, when viewing the trace in your web browser, there might be a save button that let you export the resulting TEF file?

**Comment by ASoldo** [comment]
I think go cli is preparing proper json from binary and no, it spins a local server to preview data.
https://imgur.com/a/tHQAbqf <-- this is what it looks like.

**Comment by omaskery** [comment]
I haven't worked on this project in quite a while, so it took me a while to refresh my memory. Now, looking again at the TEF document I linked previously, I am now pretty confident that I am following the specification correctly with my implementation and documentation example.

My documented example code, and therefore by extension the code you've written in this issue, generates a **thread-scoped** instant event (as denoted by the key `"s"` being set to `"t"`). Instant events that are thread-scoped are indeed allowed to have a `"stack"` key. This is therefore an issue with the chrome/catapult trace viewer not implementing the entire TEF spec.

I have provided a workaround for you:
```diff
 func tracer() {
 	t, _ := trace.TraceToFile("some.json")
 	defer t.Close()
 
 	defer t.BeginDuration("my event", trace.WithCategories("cool", "categories")).End()
 
 	// your amazing code
-	t.Instant("wow a thing happened!", trace.WithStackTrace())
+	t.Instant("wow a thing happened!")
}
```

If that's good enough for you, or Go's `runtime/trace` package has solved your needs, I will close the issue.

---

Otherwise: this library would need extending to support the `"sf"` key instead of the `"stack"` key. However, the `"sf"` key isn't compatible with TEF's "JSON Array Format" (which is the default format for streaming output in this library) since it needs to index into a file-global stack frame table only supported in TEF's "JSON Object Format" files.

Whilst this library supports writing the object format files, their downside is that they can't practically be streamed out, they effectively need to be serialised in one go. This can be problematic when tracing an application, as the application may crash without writing anything to disk, unlike streaming the array format files which would be at least partially written to disk.

If you want to press on and use TEF's JSON Object Format, then the library will need to be updated with the ability to extract stack traces from `"stack"` keys on events and move them into the stack frame table, and turn them into `"sf"` keys. It is, however, not 100% clear to me where some of the data for a stack frame should come from (the category property, though perhaps it is optional?).

In this case, I am unsure if I will have the time to address this at the moment - are you willing to submit a PR to support this?
